main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala

```diff
@@ -0,0 +1,1145 @@
+  // TODO ASSOC-TYPES this should actually do something
+  def expectTypeArguments(sym: Symbol, expectedTypes: List[Type], actualTypes: List[Type], actualLocs: List[SourceLocation], loc: SourceLocation)(implicit c: Context, root: KindedAst.Root, flix: Flix): Unit = {
+    expectedTypes.zip(actualTypes).zip(actualLocs).foreach {
+      case ((expectedType, actualType), loc) => expectTypeM(expectedType, actualType, loc)
+    }
+  }
+
+  /**
+    * Generates constraints unifying the given types.
+    */
+  def unifyType3M(tpe1: Type, tpe2: Type, tpe3: Type, loc: SourceLocation)(implicit c: Context): Unit = {
+    unifyTypeM(tpe1, tpe2, loc)
+    unifyTypeM(tpe1, tpe3, loc)
+  }
+
+  /**
+    * Generates constraints unifying the given effects.
+    */
+  // TODO ASSOC-TYPES this should actually do something
+  def unifyEffM(tpe1: Type, tpe2: Type, loc: SourceLocation)(implicit c: Context, flix: Flix): Unit = {
+    unifyTypeM(tpe1, tpe2, loc)
+  }
+
+  /**
+    * Generates constraints expecting the given types to unify.
+    */
+  // TODO ASSOC-TYPES this should actually do something
+  def expectTypeM(expected: Type, actual: Type, loc: SourceLocation)(implicit c: Context): Unit = {
+    unifyTypeM(expected, actual, loc)
+  }
+
+  /**
+    * Generates constraints expecting the given types to unify, binding them to the bound type.
+    */
+  // TODO ASSOC-TYPES what does this do?
+  def expectTypeBindM(expected: Type, actual: Type, bind: Type, loc: SourceLocation)(implicit c: Context): Unit = {
+    expectTypeM(expected, actual, loc)
+    unifyTypeM(expected, bind, loc)
+  }
+
+  /**
+    * Generates constraints unifying the given Booleans.
+    */
+  // TODO ASSOC-TYPES this should actually do something
+  def unifyBoolM(tpe1: Type, tpe2: Type, loc: SourceLocation)(implicit c: Context): Unit = {
+    unifyTypeM(tpe1, tpe2, loc)
+  }
+
+  /**
+    * Adds the given class constraints to the context.
+    */
+  def addTypeConstraintsM(tconstrs0: List[Ast.TypeConstraint], loc: SourceLocation)(implicit c: Context): Unit = {
+    val tconstrs = tconstrs0.map {
+      case Ast.TypeConstraint(head, arg, _) => Constraint.Class(head.sym, arg, c.lenv, loc)
+    }
+    c.constrs.addAll(tconstrs)
+  }
+
+  /**
+    * Marks the given type variable as rigid in the context.
+    */
+  def rigidifyM(sym: Symbol.KindedTypeVarSym)(implicit c: Context): Unit = {
+    c.renv = c.renv.markRigid(sym)
+  }
+
+  /**
+    * Enters the type variable's scope in the context.
+    */
+  def enterScopeM(sym: Symbol.KindedTypeVarSym)(implicit c: Context): Unit = {
+    c.lenv = c.lenv.enterScope(sym)
+  }
+
+  /**
+    * Exits the type variable's scope in the context.
+    */
+  def exitScopeM(sym: Symbol.KindedTypeVarSym)(implicit c: Context): Unit = {
+    c.lenv = c.lenv.exitScope(sym)
+  }
+
+  /**
+    * Contains information to perform type unification in an expression.
+    */
+  case class Context(constrs: ListBuffer[Constraint], var renv: RigidityEnv, var lenv: LevelEnv)
```

Comment:
Do we want to move Context into its own class together with its unifyType methods so it becomes like its own thing? Or is it fine to have it here?

Reply:
Yes, I think we should move it.